# Task

## T5.1: Download the 'Portugal_online_retail', 'Sweden_online_retail, and 'UK_online_retail' datasets from Canvas. Apply the apriori algorithm to all datasets using three different confidence levels. Select one confidence level for each dataset that you think works better. Determine the first three most important rules for each dataset using the selected confidence level and report them in the report cell. Explain what each rule means (Completing the report cell is required) (15%).

NOTE: You should comment on your code

In [1]:
############### Write your code in this cell (If applicable) ##################

# Import necessary libraries
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [2]:
# Load the datasets
portugal = pd.read_csv('Portugal_online_retail.csv')
sweden = pd.read_csv('Sweden_online_retail.csv')
uk = pd.read_csv('UK_online_retail.csv')

C:\Users\JOEDET\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [3]:
#display first five rows
portugal.head()

InvoiceNo  10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0     536990                       0                              0   
1     537246                       0                              0   
2     537818                       0                              0   
3     537915                       0                              0   
4     538311                       0                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE POSY  \
0                                    0                          0   
1                                    0                          0   
2                                    0                          0   
3                                    0                          0   
4                                    0                          0   

   12 PENCILS TALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE WOODLAND  \
0                                   0                              0   
1                                   0                              0   
2                                   0                              0   
3                                   0                              0   
4                                   0                              0   

   20 DOLLY PEGS RETROSPOT  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   3 STRIPEY MICE FELTCRAFT  ...  WRAP FLOWER SHOP  WRAP GINGHAM ROSE  \
0                         0  ...                 0                  0   
1                         1  ...                 0                  0   
2                         0  ...                 0                  0   
3                         0  ...                 0                  0   
4                         0  ...                 0                  0   

   WRAP PAISLEY PARK  WRAP PINK FAIRY CAKES  WRAP RED APPLES  \
0                  0                      0                0   
1                  0                      0                0   
2                  0                      0                0   
3                  0                      0                0   
4                  0                      0                0   

   WRAP RED VINTAGE DOILY  WRAP SUKI AND FRIENDS  \
0                       0                      0   
1                       0                      0   
2                       0                      0   
3                       0                      0   
4                       0                      0   

   YOU'RE CONFUSING ME METAL SIGN  ZINC FOLKART SLEIGH BELLS  \
0                               0                          0   
1                               0                          0   
2                               0                          0   
3                               0                          0   
4                               0                          0   

   ZINC WIRE KITCHEN ORGANISER  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0  

[5 rows x 714 columns]

In [4]:
#display first five rows
sweden.head()

InvoiceNo  12 PENCIL SMALL TUBE WOODLAND  \
0     538848                              0   
1     539338                              0   
2     540040                              0   
3     542428                              0   
4     542911                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
0                                    0                            0   
1                                    0                            0   
2                                    0                            0   
3                                    0                            0   
4                                    0                            0   

   12 PENCILS TALL TUBE WOODLAND  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                              0                                   0   
1                              0                                   0   
2                              0                                   0   
3                              0                                   0   
4                              0                                   1   

   3 RAFFIA RIBBONS 50'S CHRISTMAS  3 RAFFIA RIBBONS VINTAGE CHRISTMAS  \
0                                0                                   0   
1                                0                                   0   
2                                0                                   0   
3                                0                                   0   
4                                0                                   0   

   3 TIER CAKE TIN RED AND CREAM  3 TRADITIONAl BISCUIT CUTTERS  SET  ...  \
0                              0                                   0  ...   
1                              0                                   0  ...   
2                              0                                   0  ...   
3                              1                                   0  ...   
4                              0                                   0  ...   

   WOODEN STAR CHRISTMAS SCANDINAVIAN  WOODEN TREE CHRISTMAS SCANDINAVIAN  \
0                                   0                                   0   
1                                   0                                   0   
2                                   0                                   0   
3                                   0                                   0   
4                                   0                                   0   

   WOODLAND CHARLOTTE BAG  WOODLAND SMALL RED FELT HEART  \
0                       0                              0   
1                       0                              0   
2                       0                              0   
3                       0                              0   
4                       0                              0   

   WORLD WAR 2 GLIDERS ASSTD DESIGNS  WRAP  VINTAGE DOILY  \
0                                  0                    0   
1                                  1                    0   
2                                  0                    0   
3                                  0                    0   
4                                  0                    0   

   WRAP ALPHABET DESIGN  WRAP DOLLY GIRL  WRAP RED VINTAGE DOILY  \
0                     0                0                       0   
1                     0                0                       0   
2                     0                0                       0   
3                     0                0                       0   
4                     1                0                       0   

   ZINC WILLIE WINKIE  CANDLE STICK  
0                                 0  
1                                 0  
2                                 0  
3                                 0  
4                                 0  

[5 rows x 262 columns]

In [5]:
#display first five rows
uk.head()

InvoiceNo  *Boombox Ipod Classic  *USB Office Mirror Ball  \
0    536365                      0                        0   
1    536366                      0                        0   
2    536367                      0                        0   
3    536368                      0                        0   
4    536369                      0                        0   

   10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
0                       0                           0   
1                       0                           0   
2                       0                           0   
3                       0                           0   
4                       0                           0   

   12 DAISY PEGS IN WOOD BOX  12 EGG HOUSE PAINTED WOOD  \
0                          0                          0   
1                          0                          0   
2                          0                          0   
3                          0                          0   
4                          0                          0   

   12 HANGING EGGS HAND PAINTED  12 IVORY ROSE PEG PLACE SETTINGS  \
0                             0                                 0   
1                             0                                 0   
2                             0                                 0   
3                             0                                 0   
4                             0                                 0   

   12 MESSAGE CARDS WITH ENVELOPES  ...  wrongly coded 20713  \
0                                0  ...                    0   
1                                0  ...                    0   
2                                0  ...                    0   
3                                0  ...                    0   
4                                0  ...                    0   

   wrongly coded 23343  wrongly coded-23343  wrongly marked  \
0                    0                    0               0   
1                    0                    0               0   
2                    0                    0               0   
3                    0                    0               0   
4                    0                    0               0   

   wrongly marked 23343  wrongly marked carton 22804  \
0                     0                            0   
1                     0                            0   
2                     0                            0   
3                     0                            0   
4                     0                            0   

   wrongly marked. 23343 in box  wrongly sold (22719) barcode  \
0                             0                             0   
1                             0                             0   
2                             0                             0   
3                             0                             0   
4                             0                             0   

   wrongly sold as sets  wrongly sold sets  
0                     0                  0  
1                     0                  0  
2                     0                  0  
3                     0                  0  
4                     0                  0  

[5 rows x 4176 columns]

>The invoice column won't be needed in the association as we proceed, hence we will drop it in each dataset

In [6]:
#drop the invoice column
portugal.drop(columns='InvoiceNo',inplace = True)
sweden.drop(columns='InvoiceNo',inplace = True)
uk.drop(columns='InvoiceNo',inplace = True)

# Portugal Dataset

In [7]:
#view the dataset after dropping inovice column
portugal.head()

10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0                       0                              0   
1                       0                              0   
2                       0                              0   
3                       0                              0   
4                       0                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE POSY  \
0                                    0                          0   
1                                    0                          0   
2                                    0                          0   
3                                    0                          0   
4                                    0                          0   

   12 PENCILS TALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE WOODLAND  \
0                                   0                              0   
1                                   0                              0   
2                                   0                              0   
3                                   0                              0   
4                                   0                              0   

   20 DOLLY PEGS RETROSPOT  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   3 STRIPEY MICE FELTCRAFT  36 FOIL HEART CAKE CASES  ...  WRAP FLOWER SHOP  \
0                         0                         0  ...                 0   
1                         1                         0  ...                 0   
2                         0                         0  ...                 0   
3                         0                         0  ...                 0   
4                         0                         0  ...                 0   

   WRAP GINGHAM ROSE  WRAP PAISLEY PARK  WRAP PINK FAIRY CAKES  \
0                  0                  0                      0   
1                  0                  0                      0   
2                  0                  0                      0   
3                  0                  0                      0   
4                  0                  0                      0   

   WRAP RED APPLES  WRAP RED VINTAGE DOILY  WRAP SUKI AND FRIENDS  \
0                0                       0                      0   
1                0                       0                      0   
2                0                       0                      0   
3                0                       0                      0   
4                0                       0                      0   

   YOU'RE CONFUSING ME METAL SIGN  ZINC FOLKART SLEIGH BELLS  \
0                               0                          0   
1                               0                          0   
2                               0                          0   
3                               0                          0   
4                               0                          0   

   ZINC WIRE KITCHEN ORGANISER  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0  

[5 rows x 713 columns]

In [8]:
# Generate frequent itemsets from the Portugal dataset using a minimum support threshold of 10%. The output will include itemsets that are present in at least 10% of the transactions.

frq_items = apriori(portugal, min_support=0.1, use_colnames=True)

C:\Users\JOEDET\Anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [9]:
#display the frq_items table

frq_items.head()

support                          itemsets
0  0.120690     (4 TRADITIONAL SPINNING TOPS)
1  0.206897    (BAKING SET 9 PIECE RETROSPOT)
2  0.172414       (CHARLOTTE BAG SUKI DESIGN)
3  0.103448  (CLOTHES PEGS RETROSPOT PACK 24)
4  0.103448                 (DOORMAT AIRMAIL)

In [10]:
#sort the dataset with the support column 

frq_items.sort_values(by=['support'],ascending=False)

support                                           itemsets
25   0.517241                                          (POSTAGE)
30   0.241379                  (RETROSPOT TEA SET CERAMIC 11 PC)
14   0.241379                          (LUNCH BAG RED RETROSPOT)
1    0.206897                     (BAKING SET 9 PIECE RETROSPOT)
11   0.206897                              (LUNCH BAG CARS BLUE)
..        ...                                                ...
72   0.103448  (PLASTERS IN TIN VINTAGE PAISLEY, PLASTERS IN ...
73   0.103448  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...
22   0.103448                        (PLASTERS IN TIN STRONGMAN)
75   0.103448  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...
109  0.103448  (LUNCH BAG CARS BLUE, JUMBO SHOPPER VINTAGE RE...

[110 rows x 2 columns]

In [11]:
# Generate association rules based on confidence level of 70%

rules = association_rules(frq_items, metric="confidence", min_threshold=0.7)

In [12]:
rules

antecedents  \
0                       (BAKING SET 9 PIECE RETROSPOT)   
1                          (CHARLOTTE BAG SUKI DESIGN)   
2                        (RED RETROSPOT CHARLOTTE BAG)   
3                            (JUMBO BAG RED RETROSPOT)   
4                     (JUMBO BAG PINK VINTAGE PAISLEY)   
..                                                 ...   
204  (LUNCH BAG CARS BLUE, JUMBO SHOPPER VINTAGE RE...   
205  (LUNCH BAG CARS BLUE, JUMBO BAG PINK VINTAGE P...   
206     (LUNCH BAG CARS BLUE, LUNCH BAG RED RETROSPOT)   
207  (LUNCH BAG RED RETROSPOT, JUMBO SHOPPER VINTAG...   
208  (LUNCH BAG RED RETROSPOT, JUMBO BAG PINK VINTA...   

                                           consequents  antecedent support  \
0                    (RETROSPOT TEA SET CERAMIC 11 PC)            0.206897   
1                        (RED RETROSPOT CHARLOTTE BAG)            0.172414   
2                          (CHARLOTTE BAG SUKI DESIGN)            0.172414   
3                     (JUMBO BAG PINK VINTAGE PAISLEY)            0.137931   
4                            (JUMBO BAG RED RETROSPOT)            0.155172   
..                                                 ...                 ...   
204  (LUNCH BAG RED RETROSPOT, JUMBO BAG PINK VINTA...            0.137931   
205  (LUNCH BAG RED RETROSPOT, JUMBO SHOPPER VINTAG...            0.120690   
206  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...            0.137931   
207  (LUNCH BAG CARS BLUE, JUMBO BAG PINK VINTAGE P...            0.137931   
208  (LUNCH BAG CARS BLUE, JUMBO SHOPPER VINTAGE RE...            0.137931   

     consequent support   support  confidence      lift  leverage  conviction  \
0              0.241379  0.155172    0.750000  3.107143  0.105232    3.034483   
1              0.172414  0.120690    0.700000  4.060000  0.090963    2.758621   
2              0.172414  0.120690    0.700000  4.060000  0.090963    2.758621   
3              0.155172  0.120690    0.875000  5.638889  0.099287    6.758621   
4              0.137931  0.120690    0.777778  5.638889  0.099287    3.879310   
..                  ...       ...         ...       ...       ...         ...   
204            0.137931  0.103448    0.750000  5.437500  0.084423    3.448276   
205            0.137931  0.103448    0.857143  6.214286  0.086801    6.034483   
206            0.155172  0.103448    0.750000  4.833333  0.082045    3.379310   
207            0.120690  0.103448    0.750000  6.214286  0.086801    3.517241   
208            0.137931  0.103448    0.750000  5.437500  0.084423    3.448276   

     zhangs_metric  
0         0.855072  
1         0.910714  
2         0.910714  
3         0.954286  
4         0.973761  
..             ...  
204       0.946667  
205       0.954248  
206       0.920000  
207       0.973333  
208       0.946667  

[209 rows x 10 columns]

In [13]:
# Select the first three most important rules

top_three_rules = rules.head(3)

In [14]:
top_three_rules

antecedents                        consequents  \
0  (BAKING SET 9 PIECE RETROSPOT)  (RETROSPOT TEA SET CERAMIC 11 PC)   
1     (CHARLOTTE BAG SUKI DESIGN)      (RED RETROSPOT CHARLOTTE BAG)   
2   (RED RETROSPOT CHARLOTTE BAG)        (CHARLOTTE BAG SUKI DESIGN)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.206897            0.241379  0.155172        0.75  3.107143   
1            0.172414            0.172414  0.120690        0.70  4.060000   
2            0.172414            0.172414  0.120690        0.70  4.060000   

   leverage  conviction  zhangs_metric  
0  0.105232    3.034483       0.855072  
1  0.090963    2.758621       0.910714  
2  0.090963    2.758621       0.910714

# TOP 3 RULES EXPLAINED
Rule 1: (BAKING SET 9 PIECE RETROSPOT) ➡ (RETROSPOT TEA SET CERAMIC 11 PC)

- People who buy the 9-piece Retrospot baking set often also buy the 11-piece Retrospot tea set.
- This suggests they might like having a matching kitchenware set.
- Stores could put these items together on shelves or recommend them online to make buying both easier.

Rule 2: (CHARLOTTE BAG SUKI DESIGN) ➡ (RED RETROSPOT CHARLOTTE BAG)

- People who buy the Charlotte Bag Suki Design also tend to buy the Red Retrospot Charlotte Bag.
- This means they probably like the Charlotte bag design a lot.
- Stores could offer deals where you get both bags together or target ads for both bags to people who bought one already.

Rule 3: (RED RETROSPOT CHARLOTTE BAG) ➡ (CHARLOTTE BAG SUKI DESIGN)

This rule is similar to rule 2 but the other way round.
- People who buy the red Charlotte bag might also want the Suki design one.
- This again shows customers like the Charlotte bag design and might collect both colors.
- Stores could promote these bags together to highlight that they come in different colors.

# Sweden Dataset

> we will go on to analyse the sweden and uk dataset in a function at once

In [15]:
#view the dataset after dropping inovice column
sweden.head()

12 PENCIL SMALL TUBE WOODLAND  12 PENCILS SMALL TUBE RED RETROSPOT  \
0                              0                                    0   
1                              0                                    0   
2                              0                                    0   
3                              0                                    0   
4                              0                                    0   

   12 PENCILS SMALL TUBE SKULL  12 PENCILS TALL TUBE WOODLAND  \
0                            0                              0   
1                            0                              0   
2                            0                              0   
3                            0                              0   
4                            0                              0   

   3 PIECE SPACEBOY COOKIE CUTTER SET  3 RAFFIA RIBBONS 50'S CHRISTMAS  \
0                                   0                                0   
1                                   0                                0   
2                                   0                                0   
3                                   0                                0   
4                                   1                                0   

   3 RAFFIA RIBBONS VINTAGE CHRISTMAS  3 TIER CAKE TIN RED AND CREAM  \
0                                   0                              0   
1                                   0                              0   
2                                   0                              0   
3                                   0                              1   
4                                   0                              0   

   3 TRADITIONAl BISCUIT CUTTERS  SET  36 DOILIES DOLLY GIRL  ...  \
0                                   0                      0  ...   
1                                   0                      1  ...   
2                                   0                      0  ...   
3                                   0                      0  ...   
4                                   0                      0  ...   

   WOODEN STAR CHRISTMAS SCANDINAVIAN  WOODEN TREE CHRISTMAS SCANDINAVIAN  \
0                                   0                                   0   
1                                   0                                   0   
2                                   0                                   0   
3                                   0                                   0   
4                                   0                                   0   

   WOODLAND CHARLOTTE BAG  WOODLAND SMALL RED FELT HEART  \
0                       0                              0   
1                       0                              0   
2                       0                              0   
3                       0                              0   
4                       0                              0   

   WORLD WAR 2 GLIDERS ASSTD DESIGNS  WRAP  VINTAGE DOILY  \
0                                  0                    0   
1                                  1                    0   
2                                  0                    0   
3                                  0                    0   
4                                  0                    0   

   WRAP ALPHABET DESIGN  WRAP DOLLY GIRL  WRAP RED VINTAGE DOILY  \
0                     0                0                       0   
1                     0                0                       0   
2                     0                0                       0   
3                     0                0                       0   
4                     1                0                       0   

   ZINC WILLIE WINKIE  CANDLE STICK  
0                                 0  
1                                 0  
2                                 0  
3                                 0  
4                                 0  

[5 rows x 261 columns]

In [16]:
# Generate frequent itemsets from the Sweden dataset using a minimum support threshold of 10%. The output will include itemsets that are present in at least 10% of the transactions.

frq_items = apriori(sweden, min_support=0.1, use_colnames=True)

C:\Users\JOEDET\Anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [17]:
#display the frq_items table

frq_items.head()

support                           itemsets
0  0.111111  (60 CAKE CASES DOLLY GIRL DESIGN)
1  0.111111  (60 CAKE CASES VINTAGE CHRISTMAS)
2  0.138889                (GUMBALL COAT RACK)
3  0.138889           (JAM MAKING SET PRINTED)
4  0.166667           (MINI PAINT SET VINTAGE)

In [18]:
#sort the dataset with the support column 

frq_items.sort_values(by=['support'],ascending=False)

support                                           itemsets
8   0.611111                                          (POSTAGE)
14  0.194444                 (SET OF 3 CAKE TINS PANTRY DESIGN)
4   0.166667                           (MINI PAINT SET VINTAGE)
2   0.138889                                (GUMBALL COAT RACK)
3   0.138889                           (JAM MAKING SET PRINTED)
23  0.138889           (POSTAGE, RED TOADSTOOL LED NIGHT LIGHT)
20  0.138889                       (POSTAGE, GUMBALL COAT RACK)
10  0.138889                    (RED TOADSTOOL LED NIGHT LIGHT)
19  0.138889                (WORLD WAR 2 GLIDERS ASSTD DESIGNS)
0   0.111111                  (60 CAKE CASES DOLLY GIRL DESIGN)
16  0.111111                              (WALL TIDY RETROSPOT)
22  0.111111                      (POSTAGE, RABBIT NIGHT LIGHT)
21  0.111111  (PACK OF 72 RETROSPOT CAKE CASES, PACK OF 60 S...
18  0.111111                        (WOODEN OWLS LIGHT GARLAND)
17  0.111111                           (WOODEN BOX OF DOMINOES)
12  0.111111                (ROUND SNACK BOXES SET OF 4 FRUITS)
15  0.111111               (SET OF 72 RETROSPOT PAPER  DOILIES)
13  0.111111                (SET OF 12 FAIRY CAKE BAKING CASES)
1   0.111111                  (60 CAKE CASES VINTAGE CHRISTMAS)
11  0.111111                  (RETROSPOT TEA SET CERAMIC 11 PC)
9   0.111111                               (RABBIT NIGHT LIGHT)
7   0.111111                  (PACK OF 72 RETROSPOT CAKE CASES)
6   0.111111                   (PACK OF 60 SPACEBOY CAKE CASES)
5   0.111111               (PACK OF 60 PINK PAISLEY CAKE CASES)
24  0.111111                     (POSTAGE, WALL TIDY RETROSPOT)

In [19]:
# Generate association rules based on confidence level of 60%

rules = association_rules(frq_items, metric="confidence", min_threshold=0.6)

In [20]:
rules

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
2   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   
3               (RABBIT NIGHT LIGHT)                          (POSTAGE)   
4    (RED TOADSTOOL LED NIGHT LIGHT)                          (POSTAGE)   
5              (WALL TIDY RETROSPOT)                          (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889         1.0  1.636364   
1            0.111111            0.111111  0.111111         1.0  9.000000   
2            0.111111            0.111111  0.111111         1.0  9.000000   
3            0.111111            0.611111  0.111111         1.0  1.636364   
4            0.138889            0.611111  0.138889         1.0  1.636364   
5            0.111111            0.611111  0.111111         1.0  1.636364   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.098765         inf       1.000000  
2  0.098765         inf       1.000000  
3  0.043210         inf       0.437500  
4  0.054012         inf       0.451613  
5  0.043210         inf       0.437500

In [21]:
# Select the first three most important rules

top_three_rules = rules.head(3)

In [22]:
top_three_rules

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
2   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889         1.0  1.636364   
1            0.111111            0.111111  0.111111         1.0  9.000000   
2            0.111111            0.111111  0.111111         1.0  9.000000   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.098765         inf       1.000000  
2  0.098765         inf       1.000000

# TOP 3 RULES EXPLAINED
Rule 1: (GUMBALL COAT RACK) ➡ (POSTAGE)

>People who buy the Gumball Coat Rack often also purchase Postage. This rule has a high confidence of 1.0, indicating that >every transaction involving the Gumball Coat Rack also includes Postage. The lift value of 1.64 indicates that the presence of >the Gumball Coat Rack slightly increases the likelihood of Postage being purchased. This rule might not provide actionable >insights as it appears to be a common combination.

Rule 2: (PACK OF 60 SPACEBOY CAKE CASES) ➡ (PACK OF 72 RETROSPOT CAKE CASES)

>Customers who buy the Pack of 60 Spaceboy Cake Cases also tend to buy the Pack of 72 Retrospot Cake Cases. With a confidence >of 1.0, every transaction with the Pack of 60 Spaceboy Cake Cases includes the Pack of 72 Retrospot Cake Cases. The lift value >of 9.0 indicates a strong association between these two items, suggesting that customers prefer purchasing both sets of cake >cases together. Stores could consider bundling or promoting these items together.

Rule 3: (PACK OF 72 RETROSPOT CAKE CASES) ➡ (PACK OF 60 SPACEBOY CAKE CASES)

>Similar to Rule 2 but in the opposite direction, customers who buy the Pack of 72 Retrospot Cake Cases are very likely to buy >the Pack of 60 Spaceboy Cake Cases as well. This reciprocal association suggests that customers who show interest in one set >of cake cases are highly likely to also be interested in the other. Stores could use this information to cross-promote these >cake case sets.

# UK Dataset

In [23]:
#view the dataset after dropping inovice column
uk.head()

*Boombox Ipod Classic  *USB Office Mirror Ball  10 COLOUR SPACEBOY PEN  \
0                      0                        0                       0   
1                      0                        0                       0   
2                      0                        0                       0   
3                      0                        0                       0   
4                      0                        0                       0   

   12 COLOURED PARTY BALLOONS  12 DAISY PEGS IN WOOD BOX  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   

   12 EGG HOUSE PAINTED WOOD  12 HANGING EGGS HAND PAINTED  \
0                          0                             0   
1                          0                             0   
2                          0                             0   
3                          0                             0   
4                          0                             0   

   12 IVORY ROSE PEG PLACE SETTINGS  12 MESSAGE CARDS WITH ENVELOPES  \
0                                 0                                0   
1                                 0                                0   
2                                 0                                0   
3                                 0                                0   
4                                 0                                0   

   12 PENCIL SMALL TUBE WOODLAND  ...  wrongly coded 20713  \
0                              0  ...                    0   
1                              0  ...                    0   
2                              0  ...                    0   
3                              0  ...                    0   
4                              0  ...                    0   

   wrongly coded 23343  wrongly coded-23343  wrongly marked  \
0                    0                    0               0   
1                    0                    0               0   
2                    0                    0               0   
3                    0                    0               0   
4                    0                    0               0   

   wrongly marked 23343  wrongly marked carton 22804  \
0                     0                            0   
1                     0                            0   
2                     0                            0   
3                     0                            0   
4                     0                            0   

   wrongly marked. 23343 in box  wrongly sold (22719) barcode  \
0                             0                             0   
1                             0                             0   
2                             0                             0   
3                             0                             0   
4                             0                             0   

   wrongly sold as sets  wrongly sold sets  
0                     0                  0  
1                     0                  0  
2                     0                  0  
3                     0                  0  
4                     0                  0  

[5 rows x 4175 columns]

In [24]:
# Generate frequent itemsets from the UK dataset using a minimum support threshold of 10%. The output will include itemsets that are present in at least 10% of the transactions.

frq_items = apriori(uk, min_support=0.05, use_colnames=True)

C:\Users\JOEDET\Anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [25]:
#display the frq_items table

frq_items.head()

support                           itemsets
0  0.073445    (ASSORTED COLOUR BIRD ORNAMENT)
1  0.050035  (GREEN REGENCY TEACUP AND SAUCER)
2  0.062356            (HEART OF WICKER SMALL)
3  0.053999           (JAM MAKING SET PRINTED)
4  0.053517         (JAM MAKING SET WITH JARS)

In [26]:
#sort the dataset with the support column 

frq_items.sort_values(by=['support'],ascending=False)

support                              itemsets
26  0.116034  (WHITE HANGING HEART T-LIGHT HOLDER)
6   0.103820             (JUMBO BAG RED RETROSPOT)
20  0.090266            (REGENCY CAKESTAND 3 TIER)
18  0.085391                       (PARTY BUNTING)
12  0.074570             (LUNCH BAG RED RETROSPOT)
0   0.073445       (ASSORTED COLOUR BIRD ORNAMENT)
22  0.066481    (SET OF 3 CAKE TINS PANTRY DESIGN)
15  0.065302      (NATURAL SLATE HEART CHALKBOARD)
9   0.065142             (LUNCH BAG  BLACK SKULL.)
2   0.062356               (HEART OF WICKER SMALL)
5   0.062088             (JUMBO BAG PINK POLKADOT)
7   0.060695   (JUMBO SHOPPER VINTAGE RED PAISLEY)
8   0.060535              (JUMBO STORAGE BAG SUKI)
16  0.060481     (PACK OF 72 RETROSPOT CAKE CASES)
17  0.060267      (PAPER CHAIN KIT 50'S CHRISTMAS)
28  0.057642   (WOODEN PICTURE FRAME WHITE FINISH)
10  0.057053                 (LUNCH BAG CARS BLUE)
24  0.055767                      (SPOTTY BUNTING)
13  0.055606           (LUNCH BAG SPACEBOY DESIGN)
19  0.055338     (RECIPE BOX PANTRY YELLOW DESIGN)
14  0.054749               (LUNCH BAG SUKI DESIGN)
3   0.053999              (JAM MAKING SET PRINTED)
11  0.053892             (LUNCH BAG PINK POLKADOT)
25  0.053678     (VICTORIAN GLASS HANGING T-LIGHT)
23  0.053570        (SET OF 4 PANTRY JELLY MOULDS)
4   0.053517            (JAM MAKING SET WITH JARS)
21  0.051267     (ROSES REGENCY TEACUP AND SAUCER)
27  0.050463          (WOODEN FRAME ANTIQUE WHITE)
1   0.050035     (GREEN REGENCY TEACUP AND SAUCER)

In [27]:
# Generate association rules based on confidence level of 70%

rules = association_rules(frq_items, metric="confidence", min_threshold=0.01)

In [28]:
rules

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction, zhangs_metric]
Index: []

In [29]:
# Select the first three most important rules

top_three_rules = rules.head(3)

In [30]:
top_three_rules

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction, zhangs_metric]
Index: []

>No prior knowledge to why this is so as this dataset doesn't generate rules even after tweaking the **min_threshold** to different values, as thus the top three rules can't be determined.

######################## Write your report in this cell (if applicable) #############################

Note: Double click to write